{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tacco as tc\n",
    "import anndata\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "z_stack = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "andata_sp = anndata.read(f'/media/gambino/students_workdir/ilias/data/images/tile40/transcripts/tile_40_gt_z{z_stack}_expression_matrix.h5ad')\n",
    "andata_sc = anndata.read_text('/media/gambino/students_workdir/ilias/data/colorectal_cancer_sc/GSE132465_GEO_processed_CRC_10X/GSE132465_GEO_processed_CRC_10X_raw_UMI_count_matrix.txt.gz').transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_annotation_sc = pd.read_table('/media/gambino/students_workdir/ilias/data/colorectal_cancer_sc/GSE132465_GEO_processed_CRC_10X/GSE132465_GEO_processed_CRC_10X_cell_annotation.txt.gz',index_col=0)\n",
    "andata_sc.obs = cell_annotation_sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 12244 × 500\n",
       "    obs: 'mask_label', 'bbox', 'centroid-0', 'centroid-1', 'axis_major_length', 'axis_minor_length', 'area'\n",
       "    uns: 'image', 'img_files', 'masks', 'transcript_coordinates'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "andata_sp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_cells(andata_sp, min_genes=30)\n",
    "sc.pp.filter_genes(andata_sp, min_cells=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 12135 × 500\n",
       "    obs: 'mask_label', 'bbox', 'centroid-0', 'centroid-1', 'axis_major_length', 'axis_minor_length', 'area', 'n_genes'\n",
       "    var: 'n_cells'\n",
       "    uns: 'image', 'img_files', 'masks', 'transcript_coordinates'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "andata_sp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "if np.any(andata_sp.var_names.str.startswith('MT-')):\n",
    "    andata_sp.var['mt'] = andata_sp.var_names.str.startswith('MT-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "    sc.pp.calculate_qc_metrics(andata_sp, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "else:\n",
    "    sc.pp.calculate_qc_metrics(andata_sp, percent_top=None, log1p=False, inplace=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TACCO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting preprocessing\n",
      "Finished preprocessing in 1.25 seconds.\n",
      "Starting annotation of data with shape (12135, 497) and a reference of shape (63688, 497) using the following wrapped method:\n",
      "+- platform normalization: platform_iterations=0, gene_keys=Cell_type, normalize_to=adata\n",
      "   +- bisection boost: bisections=4, bisection_divisor=3\n",
      "      +- core: method=OT annotation_prior=None\n",
      "mean,std( rescaling(gene) )  2.669033144591657 15.66134505752349\n",
      "bisection run on 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/media/gambino/students_workdir/ilias/miniconda3/envs/TACCO_env/lib/python3.10/site-packages/tacco/tools/_annotate.py:177: FutureWarning: X.dtype being converted to np.float32 from uint32. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\n",
      "  adata = ad.AnnData(adata.X.copy(), obs=adata.obs[[]], var=adata.var[[]])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bisection run on 0.6666666666666667\n",
      "bisection run on 0.4444444444444444\n",
      "bisection run on 0.2962962962962963\n",
      "bisection run on 0.19753086419753085\n",
      "bisection run on 0.09876543209876543\n",
      "Finished annotation in 1.62 seconds.\n",
      "Starting preprocessing\n",
      "Annotation profiles were not found in `reference.varm[\"Class\"]`. Constructing reference profiles with `tacco.preprocessing.construct_reference_profiles` and default arguments...\n",
      "Finished preprocessing in 1.48 seconds.\n",
      "Starting annotation of data with shape (12135, 497) and a reference of shape (63688, 497) using the following wrapped method:\n",
      "+- platform normalization: platform_iterations=0, gene_keys=Class, normalize_to=adata\n",
      "   +- bisection boost: bisections=4, bisection_divisor=3\n",
      "      +- core: method=OT annotation_prior=None\n",
      "mean,std( rescaling(gene) )  2.669033144591657 15.66134505752349\n",
      "bisection run on 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/media/gambino/students_workdir/ilias/miniconda3/envs/TACCO_env/lib/python3.10/site-packages/tacco/tools/_annotate.py:177: FutureWarning: X.dtype being converted to np.float32 from uint32. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\n",
      "  adata = ad.AnnData(adata.X.copy(), obs=adata.obs[[]], var=adata.var[[]])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bisection run on 0.6666666666666667\n",
      "bisection run on 0.4444444444444444\n",
      "bisection run on 0.2962962962962963\n",
      "bisection run on 0.19753086419753085\n",
      "bisection run on 0.09876543209876543\n",
      "Finished annotation in 1.63 seconds.\n"
     ]
    }
   ],
   "source": [
    "annotations_tacco_cell_type = tc.tl.annotate(andata_sp, andata_sc, annotation_key='Cell_type')\n",
    "annotations_tacco_class = tc.tl.annotate(andata_sp, andata_sc, annotation_key='Class')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "andata_sp_tacco = andata_sp.copy()\n",
    "andata_sp_tacco.obs['pred_cell_type'] = annotations_tacco_cell_type.idxmax(axis=1)\n",
    "andata_sp_tacco.obs['pred_class'] = annotations_tacco_class.idxmax(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mask_label</th>\n",
       "      <th>bbox</th>\n",
       "      <th>centroid-0</th>\n",
       "      <th>centroid-1</th>\n",
       "      <th>axis_major_length</th>\n",
       "      <th>axis_minor_length</th>\n",
       "      <th>area</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>pred_cell_type</th>\n",
       "      <th>pred_class</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7596_0_2010_93_2113</th>\n",
       "      <td>7596</td>\n",
       "      <td>0,2010,93,2113</td>\n",
       "      <td>45.507658</td>\n",
       "      <td>2057.688285</td>\n",
       "      <td>109.680729</td>\n",
       "      <td>85.044677</td>\n",
       "      <td>7247.0</td>\n",
       "      <td>165</td>\n",
       "      <td>165</td>\n",
       "      <td>546</td>\n",
       "      <td>Stromal cells</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7597_37_3330_93_3396</th>\n",
       "      <td>7597</td>\n",
       "      <td>37,3330,93,3396</td>\n",
       "      <td>61.828131</td>\n",
       "      <td>3364.045538</td>\n",
       "      <td>68.568696</td>\n",
       "      <td>51.853015</td>\n",
       "      <td>2723.0</td>\n",
       "      <td>99</td>\n",
       "      <td>99</td>\n",
       "      <td>258</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7598_37_3450_167_3552</th>\n",
       "      <td>7598</td>\n",
       "      <td>37,3450,167,3552</td>\n",
       "      <td>108.106342</td>\n",
       "      <td>3506.736172</td>\n",
       "      <td>131.445293</td>\n",
       "      <td>91.276359</td>\n",
       "      <td>8877.0</td>\n",
       "      <td>198</td>\n",
       "      <td>198</td>\n",
       "      <td>736</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7599_111_4659_167_4751</th>\n",
       "      <td>7599</td>\n",
       "      <td>111,4659,167,4751</td>\n",
       "      <td>137.323662</td>\n",
       "      <td>4699.303434</td>\n",
       "      <td>93.839303</td>\n",
       "      <td>52.468402</td>\n",
       "      <td>3757.0</td>\n",
       "      <td>155</td>\n",
       "      <td>155</td>\n",
       "      <td>502</td>\n",
       "      <td>Myeloids</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7600_157_4871_214_4937</th>\n",
       "      <td>7600</td>\n",
       "      <td>157,4871,214,4937</td>\n",
       "      <td>183.948817</td>\n",
       "      <td>4900.310585</td>\n",
       "      <td>69.310639</td>\n",
       "      <td>50.791298</td>\n",
       "      <td>2579.0</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>126</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705140_8122_6052_8197_6127</th>\n",
       "      <td>705140</td>\n",
       "      <td>8122,6052,8197,6127</td>\n",
       "      <td>8155.671742</td>\n",
       "      <td>6087.641292</td>\n",
       "      <td>77.904241</td>\n",
       "      <td>71.367987</td>\n",
       "      <td>4335.0</td>\n",
       "      <td>177</td>\n",
       "      <td>177</td>\n",
       "      <td>563</td>\n",
       "      <td>Myeloids</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705141_8131_4410_8188_4448</th>\n",
       "      <td>705141</td>\n",
       "      <td>8131,4410,8188,4448</td>\n",
       "      <td>8159.277710</td>\n",
       "      <td>4430.278928</td>\n",
       "      <td>60.143913</td>\n",
       "      <td>35.322795</td>\n",
       "      <td>1642.0</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>93</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705142_8150_4972_8215_5038</th>\n",
       "      <td>705142</td>\n",
       "      <td>8150,4972,8215,5038</td>\n",
       "      <td>8179.948357</td>\n",
       "      <td>5002.281690</td>\n",
       "      <td>75.216194</td>\n",
       "      <td>53.344501</td>\n",
       "      <td>2982.0</td>\n",
       "      <td>99</td>\n",
       "      <td>99</td>\n",
       "      <td>343</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705143_8150_4151_8243_4281</th>\n",
       "      <td>705143</td>\n",
       "      <td>8150,4151,8243,4281</td>\n",
       "      <td>8197.131278</td>\n",
       "      <td>4220.602253</td>\n",
       "      <td>140.467850</td>\n",
       "      <td>66.865402</td>\n",
       "      <td>6391.0</td>\n",
       "      <td>186</td>\n",
       "      <td>186</td>\n",
       "      <td>875</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705144_8187_4234_8280_4328</th>\n",
       "      <td>705144</td>\n",
       "      <td>8187,4234,8280,4328</td>\n",
       "      <td>8239.094270</td>\n",
       "      <td>4284.704020</td>\n",
       "      <td>123.335300</td>\n",
       "      <td>46.986630</td>\n",
       "      <td>4328.0</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>465</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12135 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            mask_label                 bbox   centroid-0  \\\n",
       "cell_id                                                                    \n",
       "7596_0_2010_93_2113               7596       0,2010,93,2113    45.507658   \n",
       "7597_37_3330_93_3396              7597      37,3330,93,3396    61.828131   \n",
       "7598_37_3450_167_3552             7598     37,3450,167,3552   108.106342   \n",
       "7599_111_4659_167_4751            7599    111,4659,167,4751   137.323662   \n",
       "7600_157_4871_214_4937            7600    157,4871,214,4937   183.948817   \n",
       "...                                ...                  ...          ...   \n",
       "705140_8122_6052_8197_6127      705140  8122,6052,8197,6127  8155.671742   \n",
       "705141_8131_4410_8188_4448      705141  8131,4410,8188,4448  8159.277710   \n",
       "705142_8150_4972_8215_5038      705142  8150,4972,8215,5038  8179.948357   \n",
       "705143_8150_4151_8243_4281      705143  8150,4151,8243,4281  8197.131278   \n",
       "705144_8187_4234_8280_4328      705144  8187,4234,8280,4328  8239.094270   \n",
       "\n",
       "                             centroid-1  axis_major_length  axis_minor_length  \\\n",
       "cell_id                                                                         \n",
       "7596_0_2010_93_2113         2057.688285         109.680729          85.044677   \n",
       "7597_37_3330_93_3396        3364.045538          68.568696          51.853015   \n",
       "7598_37_3450_167_3552       3506.736172         131.445293          91.276359   \n",
       "7599_111_4659_167_4751      4699.303434          93.839303          52.468402   \n",
       "7600_157_4871_214_4937      4900.310585          69.310639          50.791298   \n",
       "...                                 ...                ...                ...   \n",
       "705140_8122_6052_8197_6127  6087.641292          77.904241          71.367987   \n",
       "705141_8131_4410_8188_4448  4430.278928          60.143913          35.322795   \n",
       "705142_8150_4972_8215_5038  5002.281690          75.216194          53.344501   \n",
       "705143_8150_4151_8243_4281  4220.602253         140.467850          66.865402   \n",
       "705144_8187_4234_8280_4328  4284.704020         123.335300          46.986630   \n",
       "\n",
       "                              area  n_genes  n_genes_by_counts  total_counts  \\\n",
       "cell_id                                                                        \n",
       "7596_0_2010_93_2113         7247.0      165                165           546   \n",
       "7597_37_3330_93_3396        2723.0       99                 99           258   \n",
       "7598_37_3450_167_3552       8877.0      198                198           736   \n",
       "7599_111_4659_167_4751      3757.0      155                155           502   \n",
       "7600_157_4871_214_4937      2579.0       55                 55           126   \n",
       "...                            ...      ...                ...           ...   \n",
       "705140_8122_6052_8197_6127  4335.0      177                177           563   \n",
       "705141_8131_4410_8188_4448  1642.0       48                 48            93   \n",
       "705142_8150_4972_8215_5038  2982.0       99                 99           343   \n",
       "705143_8150_4151_8243_4281  6391.0      186                186           875   \n",
       "705144_8187_4234_8280_4328  4328.0      125                125           465   \n",
       "\n",
       "                              pred_cell_type pred_class  \n",
       "cell_id                                                  \n",
       "7596_0_2010_93_2113            Stromal cells      Tumor  \n",
       "7597_37_3330_93_3396        Epithelial cells      Tumor  \n",
       "7598_37_3450_167_3552       Epithelial cells      Tumor  \n",
       "7599_111_4659_167_4751              Myeloids      Tumor  \n",
       "7600_157_4871_214_4937      Epithelial cells      Tumor  \n",
       "...                                      ...        ...  \n",
       "705140_8122_6052_8197_6127          Myeloids     Normal  \n",
       "705141_8131_4410_8188_4448  Epithelial cells      Tumor  \n",
       "705142_8150_4972_8215_5038  Epithelial cells      Tumor  \n",
       "705143_8150_4151_8243_4281  Epithelial cells      Tumor  \n",
       "705144_8187_4234_8280_4328  Epithelial cells      Tumor  \n",
       "\n",
       "[12135 rows x 12 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "andata_sp_tacco.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "andata_sp_tacco.write(f'/media/gambino/students_workdir/ilias/data/images/tile40/transcripts/tile_40_gt_z{z_stack}_expression_matrix_dc_tacco.h5ad')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TANGRAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/media/gambino/students_workdir/ilias/miniconda3/envs/TACCO_env/lib/python3.10/site-packages/tacco/utils/_utils.py:902: FutureWarning: X.dtype being converted to np.float32 from uint32. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\n",
      "  adata = ad.AnnData(adata.X, var=adata.var, obs=adata.obs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/gambino/students_workdir/ilias/miniconda3/envs/tangram/bin/python\n",
      "Score: 0.344, KL reg: 0.222\n",
      "Score: 0.811, KL reg: 0.011\n",
      "Score: 0.843, KL reg: 0.006\n",
      "Score: 0.853, KL reg: 0.004\n",
      "Score: 0.858, KL reg: 0.004\n",
      "Score: 0.860, KL reg: 0.004\n",
      "Score: 0.861, KL reg: 0.003\n",
      "Score: 0.862, KL reg: 0.003\n",
      "Score: 0.863, KL reg: 0.003\n",
      "Score: 0.863, KL reg: 0.003\n",
      "\n",
      "INFO:root:497 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:497 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 497 genes and rna_count_based density_prior in cells mode...\n",
      "INFO:root:Printing scores every 100 epochs.\n",
      "INFO:root:Saving results..\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/media/gambino/students_workdir/ilias/miniconda3/envs/TACCO_env/lib/python3.10/site-packages/tacco/utils/_utils.py:902: FutureWarning: X.dtype being converted to np.float32 from uint32. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\n",
      "  adata = ad.AnnData(adata.X, var=adata.var, obs=adata.obs)\n"
     ]
    }
   ],
   "source": [
    "annotations_tangram_cell_type = tc.tl.annotate_tangram(andata_sp, andata_sc, annotation_key='Cell_type', cluster_mode = False,\n",
    "                       conda_env='/media/gambino/students_workdir/ilias/miniconda3/envs/tangram/', device='cpu')\n",
    "annotations_tangram_class = tc.tl.annotate_tangram(andata_sp, andata_sc, annotation_key='Class', cluster_mode = False,\n",
    "                       conda_env='/media/gambino/students_workdir/ilias/miniconda3/envs/tangram/', device='cpu')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "andata_sp_tangram = andata_sp.copy()\n",
    "andata_sp_tangram.obs['pred_cell_type'] = annotations_tangram_cell_type.idxmax(axis=1)\n",
    "andata_sp_tangram.obs['pred_class'] = annotations_tangram_class.idxmax(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mask_label</th>\n",
       "      <th>bbox</th>\n",
       "      <th>centroid-0</th>\n",
       "      <th>centroid-1</th>\n",
       "      <th>axis_major_length</th>\n",
       "      <th>axis_minor_length</th>\n",
       "      <th>area</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>pred_cell_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7596_0_2010_93_2113</th>\n",
       "      <td>7596</td>\n",
       "      <td>0,2010,93,2113</td>\n",
       "      <td>45.507658</td>\n",
       "      <td>2057.688285</td>\n",
       "      <td>109.680729</td>\n",
       "      <td>85.044677</td>\n",
       "      <td>7247.0</td>\n",
       "      <td>165</td>\n",
       "      <td>165</td>\n",
       "      <td>546</td>\n",
       "      <td>Stromal cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7597_37_3330_93_3396</th>\n",
       "      <td>7597</td>\n",
       "      <td>37,3330,93,3396</td>\n",
       "      <td>61.828131</td>\n",
       "      <td>3364.045538</td>\n",
       "      <td>68.568696</td>\n",
       "      <td>51.853015</td>\n",
       "      <td>2723.0</td>\n",
       "      <td>99</td>\n",
       "      <td>99</td>\n",
       "      <td>258</td>\n",
       "      <td>Stromal cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7598_37_3450_167_3552</th>\n",
       "      <td>7598</td>\n",
       "      <td>37,3450,167,3552</td>\n",
       "      <td>108.106342</td>\n",
       "      <td>3506.736172</td>\n",
       "      <td>131.445293</td>\n",
       "      <td>91.276359</td>\n",
       "      <td>8877.0</td>\n",
       "      <td>198</td>\n",
       "      <td>198</td>\n",
       "      <td>736</td>\n",
       "      <td>Epithelial cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7599_111_4659_167_4751</th>\n",
       "      <td>7599</td>\n",
       "      <td>111,4659,167,4751</td>\n",
       "      <td>137.323662</td>\n",
       "      <td>4699.303434</td>\n",
       "      <td>93.839303</td>\n",
       "      <td>52.468402</td>\n",
       "      <td>3757.0</td>\n",
       "      <td>155</td>\n",
       "      <td>155</td>\n",
       "      <td>502</td>\n",
       "      <td>Myeloids</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7600_157_4871_214_4937</th>\n",
       "      <td>7600</td>\n",
       "      <td>157,4871,214,4937</td>\n",
       "      <td>183.948817</td>\n",
       "      <td>4900.310585</td>\n",
       "      <td>69.310639</td>\n",
       "      <td>50.791298</td>\n",
       "      <td>2579.0</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>126</td>\n",
       "      <td>B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705140_8122_6052_8197_6127</th>\n",
       "      <td>705140</td>\n",
       "      <td>8122,6052,8197,6127</td>\n",
       "      <td>8155.671742</td>\n",
       "      <td>6087.641292</td>\n",
       "      <td>77.904241</td>\n",
       "      <td>71.367987</td>\n",
       "      <td>4335.0</td>\n",
       "      <td>177</td>\n",
       "      <td>177</td>\n",
       "      <td>563</td>\n",
       "      <td>Myeloids</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705141_8131_4410_8188_4448</th>\n",
       "      <td>705141</td>\n",
       "      <td>8131,4410,8188,4448</td>\n",
       "      <td>8159.277710</td>\n",
       "      <td>4430.278928</td>\n",
       "      <td>60.143913</td>\n",
       "      <td>35.322795</td>\n",
       "      <td>1642.0</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>93</td>\n",
       "      <td>T cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705142_8150_4972_8215_5038</th>\n",
       "      <td>705142</td>\n",
       "      <td>8150,4972,8215,5038</td>\n",
       "      <td>8179.948357</td>\n",
       "      <td>5002.281690</td>\n",
       "      <td>75.216194</td>\n",
       "      <td>53.344501</td>\n",
       "      <td>2982.0</td>\n",
       "      <td>99</td>\n",
       "      <td>99</td>\n",
       "      <td>343</td>\n",
       "      <td>Epithelial cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705143_8150_4151_8243_4281</th>\n",
       "      <td>705143</td>\n",
       "      <td>8150,4151,8243,4281</td>\n",
       "      <td>8197.131278</td>\n",
       "      <td>4220.602253</td>\n",
       "      <td>140.467850</td>\n",
       "      <td>66.865402</td>\n",
       "      <td>6391.0</td>\n",
       "      <td>186</td>\n",
       "      <td>186</td>\n",
       "      <td>875</td>\n",
       "      <td>Myeloids</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705144_8187_4234_8280_4328</th>\n",
       "      <td>705144</td>\n",
       "      <td>8187,4234,8280,4328</td>\n",
       "      <td>8239.094270</td>\n",
       "      <td>4284.704020</td>\n",
       "      <td>123.335300</td>\n",
       "      <td>46.986630</td>\n",
       "      <td>4328.0</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>465</td>\n",
       "      <td>Epithelial cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12135 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            mask_label                 bbox   centroid-0  \\\n",
       "cell_id                                                                    \n",
       "7596_0_2010_93_2113               7596       0,2010,93,2113    45.507658   \n",
       "7597_37_3330_93_3396              7597      37,3330,93,3396    61.828131   \n",
       "7598_37_3450_167_3552             7598     37,3450,167,3552   108.106342   \n",
       "7599_111_4659_167_4751            7599    111,4659,167,4751   137.323662   \n",
       "7600_157_4871_214_4937            7600    157,4871,214,4937   183.948817   \n",
       "...                                ...                  ...          ...   \n",
       "705140_8122_6052_8197_6127      705140  8122,6052,8197,6127  8155.671742   \n",
       "705141_8131_4410_8188_4448      705141  8131,4410,8188,4448  8159.277710   \n",
       "705142_8150_4972_8215_5038      705142  8150,4972,8215,5038  8179.948357   \n",
       "705143_8150_4151_8243_4281      705143  8150,4151,8243,4281  8197.131278   \n",
       "705144_8187_4234_8280_4328      705144  8187,4234,8280,4328  8239.094270   \n",
       "\n",
       "                             centroid-1  axis_major_length  axis_minor_length  \\\n",
       "cell_id                                                                         \n",
       "7596_0_2010_93_2113         2057.688285         109.680729          85.044677   \n",
       "7597_37_3330_93_3396        3364.045538          68.568696          51.853015   \n",
       "7598_37_3450_167_3552       3506.736172         131.445293          91.276359   \n",
       "7599_111_4659_167_4751      4699.303434          93.839303          52.468402   \n",
       "7600_157_4871_214_4937      4900.310585          69.310639          50.791298   \n",
       "...                                 ...                ...                ...   \n",
       "705140_8122_6052_8197_6127  6087.641292          77.904241          71.367987   \n",
       "705141_8131_4410_8188_4448  4430.278928          60.143913          35.322795   \n",
       "705142_8150_4972_8215_5038  5002.281690          75.216194          53.344501   \n",
       "705143_8150_4151_8243_4281  4220.602253         140.467850          66.865402   \n",
       "705144_8187_4234_8280_4328  4284.704020         123.335300          46.986630   \n",
       "\n",
       "                              area  n_genes  n_genes_by_counts  total_counts  \\\n",
       "cell_id                                                                        \n",
       "7596_0_2010_93_2113         7247.0      165                165           546   \n",
       "7597_37_3330_93_3396        2723.0       99                 99           258   \n",
       "7598_37_3450_167_3552       8877.0      198                198           736   \n",
       "7599_111_4659_167_4751      3757.0      155                155           502   \n",
       "7600_157_4871_214_4937      2579.0       55                 55           126   \n",
       "...                            ...      ...                ...           ...   \n",
       "705140_8122_6052_8197_6127  4335.0      177                177           563   \n",
       "705141_8131_4410_8188_4448  1642.0       48                 48            93   \n",
       "705142_8150_4972_8215_5038  2982.0       99                 99           343   \n",
       "705143_8150_4151_8243_4281  6391.0      186                186           875   \n",
       "705144_8187_4234_8280_4328  4328.0      125                125           465   \n",
       "\n",
       "                              pred_cell_type  \n",
       "cell_id                                       \n",
       "7596_0_2010_93_2113            Stromal cells  \n",
       "7597_37_3330_93_3396           Stromal cells  \n",
       "7598_37_3450_167_3552       Epithelial cells  \n",
       "7599_111_4659_167_4751              Myeloids  \n",
       "7600_157_4871_214_4937               B cells  \n",
       "...                                      ...  \n",
       "705140_8122_6052_8197_6127          Myeloids  \n",
       "705141_8131_4410_8188_4448           T cells  \n",
       "705142_8150_4972_8215_5038  Epithelial cells  \n",
       "705143_8150_4151_8243_4281          Myeloids  \n",
       "705144_8187_4234_8280_4328  Epithelial cells  \n",
       "\n",
       "[12135 rows x 11 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "andata_sp_tangram.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "andata_sp_tangram.write(f'/media/gambino/students_workdir/ilias/data/images/tile40/transcripts/tile_40_gt_z{z_stack}_expression_matrix_dc_tangram.h5ad')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TACCO_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
